nats-experimental/src/lib.rs

```diff
@@ -123,29 +145,88 @@ pub enum ClientOp {
     Connect(ConnectInfo),
 }
 
+trait AsyncReadWrite: AsyncWrite + AsyncRead + Send + Unpin {}
+
+impl<T> AsyncReadWrite for T where T: AsyncRead + AsyncWrite + Unpin + Send {}
+
 /// A framed connection
 ///
 /// The type will probably not be public.
 pub struct Connection {
-    stream: BufWriter<TcpStream>,
+    stream: Box<dyn AsyncReadWrite>,
     buffer: BytesMut,
 }
 
 impl Connection {
-    pub async fn connect<A: ToServerAddrs>(addrs: A) -> Result<Connection, io::Error> {
-        let a = addrs.to_server_addrs()?.into_iter().next().ok_or_else(|| {
+    pub async fn connect_with_options<A: ToServerAddrs>(
+        addrs: A,
+        options: options::Options,
+    ) -> io::Result<Connection> {
+        let addr = addrs.to_server_addrs()?.into_iter().next().ok_or_else(|| {
             io::Error::new(
                 ErrorKind::Other,
                 "did not found a single url in the url list",
             )
         })?;
-        let tcp_stream = TcpStream::connect((a.host(), a.port())).await?;
+
+        let tls_config = tls::config_tls(&options).await?;
+
+        let tcp_stream = TcpStream::connect((addr.host(), addr.port())).await?;
         tcp_stream.set_nodelay(true)?;
 
-        Ok(Connection {
-            stream: BufWriter::new(tcp_stream),
+        let mut connection = Connection {
+            stream: Box::new(BufWriter::new(tcp_stream)),
             buffer: BytesMut::new(),
-        })
+        };
+
+        let op = connection.read_op().await?;
+        let info = match op {
+            Some(ServerOp::Info(info)) => info,
+            Some(op) => {
+                return Err(io::Error::new(
+                    ErrorKind::Other,
+                    format!("expected INFO, got {:?}", op),
+                ))
+            }
+            None => {
+                return Err(io::Error::new(
+                    ErrorKind::Other,
+                    "expected INFO, got nothing",
+                ))
+            }
+        };
+
+        let tls_required = options.tls_required || info.tls_required || addr.tls_required();
+
+        if tls_required {
+            let tls_config = Arc::new(tls_config);
+            let connector = tokio_rustls::TlsConnector::try_from(tls_config).map_err(|err| {
```

Comment:
Nit, since we're using connector as a thing here too.

```suggestion
            let tls_connector = tokio_rustls::TlsConnector::try_from(tls_config).map_err(|err| {
```

Reply:
fixed.